global: Correct extern K_KERNEL_STACK_ARRAY_DEFINE and K_THREAD_STACK_DEFINE usage

stephanosio opened the pull request:
```
kernel: thread_stack: Add K_THREAD_STACK_DECLARE and its friends 

This commit adds the following new thread stack declaration macros that
replace the now-deprecated thread stack "extern" macros:

  K_KERNEL_STACK_DECLARE
    * replaces K_KERNEL_STACK_EXTERN.
    * accepts a stack size parameter.

  K_KERNEL_STACK_ARRAY_DECLARE.
    * replaces K_KERNEL_STACK_ARRAY_EXTERN.

  K_KERNEL_PINNED_STACK_ARRAY_DECLARE.
    * replaces K_KERNEL_PINNED_STACK_ARRAY_EXTERN.

  K_THREAD_STACK_DECLARE
    * replaces K_THREAD_STACK_EXTERN.
    * accepts a stack size parameter.

  K_THREAD_STACK_ARRAY_DECLARE
    * replaces K_THREAD_STACK_ARRAY_EXTERN.

The term "declare" has been chosen in place of "extern" in order to
align with the existing terminology used throughout the kernel code.

Note that the K_{KERNEL,THREAD}_STACK_DECLARE macro accepts a new stack
size parameter so that the `sizeof` operator can be used with the
external stack symbols declared using it.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

```
global: Correct extern K_KERNEL_STACK_ARRAY_DEFINE usage 

This commit corrects all `extern K_KERNEL_STACK_ARRAY_DEFINE` macro
usages to use the `K_KERNEL_STACK_ARRAY_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

```
global: Correct extern K_THREAD_STACK_DEFINE usage 

This commit corrects all `extern K_THREAD_STACK_DEFINE` macro usages
to use the `K_THREAD_STACK_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

stephanosio committed:
global: Correct extern K_KERNEL_STACK_ARRAY_DEFINE usage

This commit corrects all `extern K_KERNEL_STACK_ARRAY_DEFINE` macro
usages to use the `K_KERNEL_STACK_ARRAY_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
global: Correct extern K_THREAD_STACK_DEFINE usage

This commit corrects all `extern K_THREAD_STACK_DEFINE` macro usages
to use the `K_THREAD_STACK_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
```
/__w/zephyr/zephyr/arch/arm/core/aarch32/thread.c
In file included from ../../../../../../include/zephyr/kernel_includes.h:39,
                 from ../../../../../../include/zephyr/kernel.h:17,
                 from /__w/zephyr/zephyr/arch/arm/core/aarch32/thread.c:15:
/__w/zephyr/zephyr/arch/arm/core/aarch32/thread.c: In function 'z_arm_switch_to_main_no_multithreading':
../../../../../../include/zephyr/kernel/thread_stack.h:288:43: error: invalid application of 'sizeof' to incomplete type 'k_thread_stack_t[]' {aka 'struct z_thread_stack_element[]'}
  288 | #define K_KERNEL_STACK_SIZEOF(sym) (sizeof(sym) - K_KERNEL_STACK_RESERVED)
      |                                           ^
../../../../../../include/zephyr/kernel/thread_stack.h:298:32: note: in expansion of macro 'K_KERNEL_STACK_SIZEOF'
  298 | #define K_THREAD_STACK_SIZEOF  K_KERNEL_STACK_SIZEOF
      |                                ^~~~~~~~~~~~~~~~~~~~~
/__w/zephyr/zephyr/arch/arm/core/aarch32/thread.c:615:3: note: in expansion of macro 'K_THREAD_STACK_SIZEOF'
  615 |   K_THREAD_STACK_SIZEOF(z_main_stack);
      |   ^~~~~~~~~~~~~~~~~~~~~
```

This is a mess. We need to change the `K_THREAD_STACK_DEFINE` macro to accept the stack size parameter and define the extern-ed variable with non-empty size, as does `K_KERNEL_STACK_ARRAY_DEFINE`.

stephanosio committed:
samples: kernel: condition_variables: Remove unneeded `tstack` extern

This commit removes the `tstack` symbol extern that is not used
anywhere in this sample.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: kernel: msgq_api: Migrate to K_THREAD_STACK_DECLARE

This commit updates all deprecated `K_THREAD_STACK_EXTERN` macro usages
to use the `K_THREAD_STACK_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: kernel: schedule_api: Migrate to K_THREAD_STACK_DECLARE

This commit updates all deprecated `K_THREAD_STACK_EXTERN` macro usages
to use the `K_THREAD_STACK_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: kernel: thread_apis: Migrate to K_THREAD_STACK_DECLARE

This commit updates all deprecated `K_THREAD_STACK_EXTERN` macro usages
to use the `K_THREAD_STACK_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: benchmarks: footprint: Migrate to K_THREAD_STACK_DECLARE

This commit updates all deprecated `K_THREAD_STACK_EXTERN` macro usages
to use the `K_THREAD_STACK_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: benchmarks: sys_kernel: Migrate to K_THREAD_STACK_DECLARE

This commit updates all deprecated `K_THREAD_STACK_EXTERN` macro usages
to use the `K_THREAD_STACK_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

andyross commented:
Is it worth renaming _EXTERN() to _DECLARE() while you're here (and implementing the older one as a deprecated wrapper, I guess), as that (I think) better matches the usage in other Zephyr macro APIs?

stephanosio commented:
> Is it worth renaming _EXTERN() to _DECLARE() while you're here (and implementing the older one as a deprecated wrapper, I guess), as that (I think) better matches the usage in other Zephyr macro APIs?

That sounds very reasonable.

stephanosio committed:
kernel: thread_stack: Fix macro documentation group

This commit ensures that all thread stack API macros are included in
the `thread_stack_api` documentation group.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
kernel: thread_stack: Enforce consistent terminology

This commit updates the thread stack API documentation such the
terminology used by the documentation matches that of the actual macro
and function names.

For instance, the term "declare" in the context of "defining" a stack
has been changed to "define."

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
kernel: thread_stack: Remove redundant def command

This commit removes the `@def` command in the Doxygen documentation
because Doxygen is smart enough to figure out that a comment block is
for the following `#define`, and it is completely redundant.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
kernel: thread_stack: Add K_THREAD_STACK_DECLARE and its friends

This commit adds the following new thread stack declaration macros that
replace the now-deprecated thread stack "extern" macros:

  K_KERNEL_STACK_DECLARE
    * replaces K_KERNEL_STACK_EXTERN.
    * accepts a stack size parameter.

  K_KERNEL_STACK_ARRAY_DECLARE.
    * replaces K_KERNEL_STACK_ARRAY_EXTERN.

  K_KERNEL_PINNED_STACK_ARRAY_DECLARE.
    * replaces K_KERNEL_PINNED_STACK_ARRAY_EXTERN.

  K_THREAD_STACK_DECLARE
    * replaces K_THREAD_STACK_EXTERN.
    * accepts a stack size parameter.

  K_THREAD_STACK_ARRAY_DECLARE
    * replaces K_THREAD_STACK_ARRAY_EXTERN.

The term "declare" has been chosen in place of "extern" in order to
align with the existing terminology used throughout the kernel code.

Note that the K_{KERNEL,THREAD}_STACK_DECLARE macro accepts a new stack
size parameter so that the `sizeof` operator can be used with the
external stack symbols declared using it.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
arch: arm: Migrate to K_KERNEL_STACK_ARRAY_DECLARE

This commit updates all deprecated `K_KERNEL_STACK_ARRAY_EXTERN` macro
usages to use the `K_KERNEL_STACK_ARRAY_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
kernel: Migrate to K_KERNEL_PINNED_STACK_ARRAY_DECLARE

This commit updates all deprecated `K_KERNEL_PINNED_STACK_ARRAY_EXTERN`
macro usages to use the `K_KERNEL_PINNED_STACK_ARRAY_DECLARE` macro
instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
tests: kernel: schedule_api: Migrate to K_THREAD_STACK_ARRAY_DECLARE

This commit updates all deprecated `K_THREAD_STACK_ARRAY_EXTERN` macro
usages to use the `K_THREAD_STACK_ARRAY_DECLARE` macro instead.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

andyross reviewed (approved):
Looks good to me
